File: cutout/pipeline.py
"""Cutout's end-to-end matte processing pipeline."""
import numpy as np
from PIL import Image
from .decontaminate import decontaminate
from .refiner import refine_alpha
from .segmenter import CutoutSegmenter


class CutoutPipeline:
    def __init__(self):
        try:
            self.segmenter = CutoutSegmenter()
            self.engine = "birefnet"
        except Exception as primary_error:
            from .u2net import U2NetSegmenter

            try:
                self.segmenter = U2NetSegmenter()
                self.engine = "u2net"
            except Exception:
                raise primary_error

    def process(self, image: Image.Image) -> Image.Image:
        alpha = self.segmenter.predict_alpha(image)
        if self.engine == "birefnet":
            alpha = refine_alpha(self.segmenter, image, alpha)
        return decontaminate(image, alpha)

File: cutout/pipeline.js
import sharp from 'sharp';
import { CutoutSegmenter } from './segmenter.js';
import { refineAlpha } from './refiner.js';
import { decontaminate } from './decontaminate.js';

export class CutoutPipeline {
  constructor(segmenter) { this.segmenter = segmenter; }

  static async create() { return new CutoutPipeline(await CutoutSegmenter.create()); }

  async process(image) {
    const alpha = await this.segmenter.predictAlpha(image);
    const refined = await refineAlpha(this.segmenter, image, alpha);
    return decontaminate(image, refined);
  }

  async processFile(inputPath, outputPath) {
    const image = await sharp(inputPath).toBuffer();
    const output = await this.process(image);
    await sharp(output).png({ compressionLevel: 9 }).toFile(outputPath);
    return outputPath;
  }
}
